torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,401 @@
+                    pipe, futureResponseMessage, err, messageId);
+              });
+        }
+      });
+}
+
+std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+    const WorkerInfo& toWorkerInfo,
+    Message&& requestMessage) {
+  TORCH_CHECK(
+      requestMessage.isRequest(),
+      "TensorPipeAgent::send(..) is only for sending requests.");
+
+  if (!isRunning_) {
+    auto err = c10::str(
+        "Node ",
+        RpcAgent::getWorkerInfo().id_,
+        "tried to send() a message of type ",
+        requestMessage.type(),
+        " but RPC is no longer running on this node.");
+    throw std::runtime_error(err);
+  }
+
+  const auto& url = findWorkerURL(toWorkerInfo);
+
+  std::unique_lock<std::mutex> lock(mutex_);
+
+  // See if we already have a connection to this address or not
+  auto it = connectedPipes_.find(toWorkerInfo.id_);
+  if (it == connectedPipes_.end()) {
+    std::tie(it, std::ignore) =
+        connectedPipes_.emplace(toWorkerInfo.id_, context_->connect(url));
+  }
+  std::shared_ptr<tensorpipe::Pipe>& pipe = it->second;
+
+  std::shared_ptr<FutureMessage> futureResponseMessage =
+      std::make_shared<FutureMessage>();
+  requestMessage.setId(nextMessageID_++);
+  pendingResponseMessage_[requestMessage.id()] = futureResponseMessage;
+
+  // Don't need to hold lock while calling tensorpipe API.
+  lock.unlock();
+
+  pipeWrite(
+      pipe,
+      std::move(requestMessage),
+      [this, &pipe, futureResponseMessage](const tensorpipe::Error& error) {
+        if (error) {
+          LOG(WARNING) << "Write callback error: " << error.what();
+          futureResponseMessage->setError(error.what());
+          return;
+        }
+
+        pipeRead(
+            pipe,
+            [this](const tensorpipe::Error& error, Message&& responseMessage) {
+              if (error) {
+                LOG(WARNING) << "Read response error: " << error.what();
+                std::lock_guard<std::mutex> lock(mutex_);
+                // We may get garbage content in responseMessage upon error.
+                // Flushing all future messages due to error state.
+                for (auto& p : pendingResponseMessage_) {
+                  uint64_t messageId = p.first;
+                  std::shared_ptr<FutureMessage>& futureMessage = p.second;
+                  futureMessage->setError(error.what());
+                }
+                pendingResponseMessage_.clear();
+                readError_ = true;
+                return;
+              }
+
+              // Identify future response message by message ID
+              uint64_t messageId = responseMessage.id();
+              std::shared_ptr<FutureMessage> futureResponseMessage;
+              {
+                std::lock_guard<std::mutex> lock(mutex_);
+                // One read error will lead all following callbacks to be
+                // invoked with error, and shouldn't read here.
+                TORCH_INTERNAL_ASSERT(!readError_, "Shouldn't in error state");
```

Comment:
I suspect this assert may fail. If one pipe gets an error, the global `readError_` will be set, but other pipes may still continue to work properly. I think that when we receive an error on a pipe we should only fail the messages being sent on that pipe, and keep the other ones untouched. This is the most robust solution, especially for elastic training scenarios where machines are expected to come and go. Alternatively, to keep it simple at first, we can proactively close all pipes when one of them returns an error.